### PR TITLE
Feature/issue 1272 gradient of hessian autodiff hpp

### PR DESCRIPTION
**Pull request by rtrangucci**
#### Summary:

Addresses #1272 

#### Intended Effect:

Allows for automatic third order derivative tests, when implemented alongside a finite difference version of grad_hessian.

#### How to Verify:

Added a unit test to src/test/unit/agrad/autodiff_test.cpp that runs grad_hessian on a normal_log density and a third-order polynomial against an analytically calculated gradient of the Hessian calculation, as well as an analytically calculated Hessian versus the Hessian returned by grad_hessian. 
```
$ ./runTests.py src/test/unit/agrad/autodiff_test.cpp
```

#### Side Effects:

None.

#### Documentation:

None in the manual. Added doxygen documentation to agrad::grad_hessian.

#### Reviewer Suggestions:

Any

### CODE REVIEW

**Comment by syclik** [comment]
@rtrangucci, can you remove tabs from the code? I haven't walked through the code yet, but cpplint spotted commas in the code.

**Comment by syclik** [comment]
Looks good. Let's get some of the cpplint warnings squashed before this gets in:
http://d1m1s1b1.stat.columbia.edu:8080/job/Stan%20Pull%20Request%20-%20CppLint/21/warnings10Result/new/

Most of it will be fixing spaces. Want help?

**Comment by bob-carpenter** [comment]
Sweet --- I'm glad to have some company as a style curmudgeon (or
OCD coder --- take your pick).

- Bob

**Comment by rtrangucci** [comment]
Sure, help would be much appreciated. I may be reading the output wrong,
but it seems like cpplint objects to template instantiation without spaces
after commas: Eigen::Matrix<T,Eigen::Dynamic,1> for instance. Do we want to
move to Eigen::Matrix<T, Eigen::Dynamic, 1> ?


Rob

**Comment by syclik** [comment]
On Mon, Feb 9, 2015 at 12:25 PM, Rob Trangucci <notifications@github.com>
wrote:

> Sure, help would be much appreciated. I may be reading the output wrong,
> but it seems like cpplint objects to template instantiation without spaces
> after commas: Eigen::Matrix<T,Eigen::Dynamic,1> for instance. Do we want to
> move to Eigen::Matrix<T, Eigen::Dynamic, 1> ?
>

Yes.

**Comment by rtrangucci** [comment]
@syclik cpplint should only throw 4 warnings now, 3 for lines over 80, and one for include what you use for a std::vector instantiation.

FYI, the vim search/replace regexs I used to fix the top three warnings:

%s/\([^" "]\)==\([^" "]\)/\1 == \2/gc  // whitespace/operators: no single spaces before and after "=="
%s/,\([^" "]\)/, \1/gc  // whitespace/commas: no single space after a comma
%s/^ \+\n/\r/gc  // whitespace/end_of_line: line ends in whitespace

File: src/stan/agrad/autodiff.hpp [hunk 1]
```diff
@@ -359,6 +358,76 @@ namespace stan {
...
+                 double& fx,
+                 Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic>& H,
+                 std::vector<Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic> >& grad_H) {
+      using Eigen::Matrix;
+      using Eigen::Dynamic;
+      start_nested();
+      fx = f(x);
+      try {
+        int d = x.size();
+        H.resize(d, d);
```
Comment (syclik) [inline src/stan/agrad/autodiff.hpp]: Why isn't `grad_H` appropriately resized?

File: src/stan/agrad/autodiff.hpp [hunk 2]
```diff
@@ -359,6 +358,76 @@ namespace stan {
...
+        for (int i = 0; i < d; ++i)
+          grad_H.push_back(Matrix<double, Dynamic, Dynamic>(d, d));
+        for (int i = 0; i < d; ++i) {
+          for (int j = i; j < d; ++j) {
+            Matrix<fvar<fvar<var> >, Dynamic, 1> x_ffvar(d);
```
Comment (syclik) [inline src/stan/agrad/autodiff.hpp]: Shouldn't the `start_nested()` be here?

File: src/stan/agrad/autodiff.hpp [hunk 3]
```diff
@@ -359,6 +358,76 @@ namespace stan {
...
+                                            fvar<var>(j == k, 0));
+            fvar<fvar<var> > fx_ffvar = f(x_ffvar);
+            H(i, j) = fx_ffvar.d_.d_.val();
+            H(j, i) = H(i, j);
+            stan::agrad::grad(fx_ffvar.d_.d_.vi_);
+            for (int k = 0; k < d; ++k) {
+              grad_H[i](j, k) = x_ffvar(k).val_.val_.adj();
+              grad_H[j](i, k) = grad_H[i](j, k);
+            }
+          }
```
Comment (syclik) [inline src/stan/agrad/autodiff.hpp]: with a `recover_memory_nested()` here.

File: src/stan/agrad/autodiff.hpp [hunk 4]
```diff
@@ -359,6 +358,76 @@ namespace stan {
...
+        }
+      } catch (const std::exception& e) {
+        stan::agrad::recover_memory_nested();
+        throw;
+      }
+      stan::agrad::recover_memory_nested();
```
Comment (syclik) [inline src/stan/agrad/autodiff.hpp]: and this `recover_memory_nested()` removed?

**Comment by rtrangucci** [comment]
@syclik I ripped all these patterns from grad_tr_mat_times_hessian, hessian, and hessian_times_vector, so let's fix all of them if they're wrong

**Comment by rtrangucci** [comment]
@syclik the changes you suggested to hessian and grad_hessian from your code review increased the speed of the hessian operation by about 30% while the grad_hessian was sped up by about 50%.

File: src/stan/agrad/autodiff.hpp (hunk 1 above)
Comment (rtrangucci) [inline src/stan/agrad/autodiff.hpp]: FYI for the future, we can do the vector resize and the Matrix construction at the same time with:

grad_H.resize(d, Eigen::Matrix<double,-1,-1>(d,d));

File: src/stan/agrad/autodiff.hpp [hunk 5]
```diff
@@ -359,6 +357,75 @@ namespace stan {
...
+    void
+    grad_hessian(const F& f,
+                 const Eigen::Matrix<double, Eigen::Dynamic, 1>& x,
+                 double& fx,
+                 Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic>& H,
+                 std::vector<Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic> >& grad_H) {
+      using Eigen::Matrix;
+      using Eigen::Dynamic;
+     	fx = f(x);
+			int d = x.size();
```
Comment (bob-carpenter) [inline src/stan/agrad/autodiff.hpp]: You've got to lose the tabs or whatever's messing up the code indentantation.  I don't know if vi can do it automatically.  I have emacs set up to never save any code files with tabs.

Comment (rtrangucci) [inline src/stan/agrad/autodiff.hpp]: I don't know where these come from, my config file for vim should be
automatically exchanging tabs for spaces. Clearly it is failing. Will fix.

On Tuesday, February 10, 2015, Bob Carpenter <notifications@github.com>
wrote:

> In src/stan/agrad/autodiff.hpp
> <https://github.com/stan-dev/stan/pull/1276#discussion_r24472702>:
>
> > +     * @param[in] x Argument to function
> > +     * @param[out] fx Function applied to argument
> > +     * @param[out] H Hessian of function at argument
> > +     * @param[out] grad_H Gradient of the Hessian of function at argument
> > +     */
> > +    template <typename F>
> > +    void
> > +    grad_hessian(const F& f,
> > +                 const Eigen::Matrix<double, Eigen::Dynamic, 1>& x,
> > +                 double& fx,
> > +                 Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic>& H,
> > +                 std::vector<Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic> >& grad_H) {
> > +      using Eigen::Matrix;
> > +      using Eigen::Dynamic;
> > +     	fx = f(x);
> > +			int d = x.size();
>
> You've got to lose the tabs or whatever's messing up the code
> indentantation. I don't know if vi can do it automatically. I have emacs
> set up to never save any code files with tabs.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/stan-dev/stan/pull/1276/files#r24472702>.
>


-- 

Rob

**Comment by syclik** [comment]
Jenkins, retest this please.

File: src/stan/agrad/autodiff.hpp [hunk 6]
```diff
@@ -359,6 +357,75 @@ namespace stan {
...
+     * requires the functions to be defined in terms of the libraries
+     * defined in Stan or in terms of functions with appropriately
+     * general namespace imports that eventually depend on functions
+     * defined in Stan.
+     *
+     * @tparam F Type of function
+     * @param[in] f Function
+     * @param[in] x Argument to function
+     * @param[out] fx Function applied to argument
+     * @param[out] H Hessian of function at argument
```
Comment (syclik) [inline src/stan/agrad/autodiff.hpp]: Doxygen is balking at the comments. Do you have oxygen installed? Type `make doxygen`

Comment (rtrangucci) [inline src/stan/agrad/autodiff.hpp]: When I run make doxygen, I get the following error:

RobTrangucci ~/stan 18:56 [
feature/issue-1272-gradient-of-hessian-autodiff_hpp*] - 593:

$ make doxygen

mkdir -p doc/api

doxygen src/doxygen/doxygen.cfg

make: doxygen: No such file or directory

make: *** [doxygen] Error 1


Rob

Comment (rtrangucci) [inline src/stan/agrad/autodiff.hpp]: Disregard, I didn't have doxygen installed. I don't think I changed these comments between when the tests passed 5 days ago, so I'm not sure what changed...

**Comment by syclik** [comment]
I'll try to hunt it down later.

Does it work with your setup? Search for warnings in your output.

**Comment by rtrangucci** [comment]
Doc should be all set.


Rob

**Comment by syclik** [comment]
Jenkins, retest this please.

**Comment by syclik** [comment]
Looks good. The last push I made wasn't substantive. We can merge once the tests come back passing.

**Pull request merged**